File: 2023/src/solutions/day_02.rs
use std::collections::HashMap;

use aoc_common::{Solution, load};
use itertools::Itertools;

pub struct Day02;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
enum Color {
    Red,
    Blue,
    Green,
}

struct Game {
    pub id: isize,
    pub reveals: Vec<HashMap<Color, isize>>
}

impl Solution for Day02 {
    fn name(&self) -> String {
        "Cube Conundrum".to_owned()
    }

    fn part_1(&self) -> String {
        let bag = HashMap::from([
            (Color::Blue, 14),
            (Color::Red, 12),
            (Color::Green, 13)
        ]);

        let input = load("day_02");
        let game = input.lines()
            .map(|line| parse_game(line))
            .collect_vec();

        let possible_games: isize = game.iter()
            .filter_map(|g| is_game_possible(g, &bag))
            .sum();

        possible_games.to_string()
    }

    fn part_2(&self) -> String {
        let input = load("day_02");
        let games = input.lines()
            .map(|line| parse_game(line))
            .collect_vec();

        let total_power: isize = games.iter()
            .map(|game| get_smallest_bag(game))
            .map(|bag| bag.values().product::<isize>())
            .sum();

        total_power.to_string()
    }
}

fn parse_game(line: &str) -> Game {
    let (game_part, reveals_part) = line.split_once(':').unwrap();
    let id: isize = game_part.chars()
        .filter(|c| c.is_ascii_digit())
        .collect::<String>()
        .parse()
        .unwrap();
    
    let reveals = reveals_part.split(';')
        .map(|r| r.split(',').map(parse_color_reveal).collect())
        .collect();

    Game { id, reveals }
}

fn parse_color_reveal(reveal: &str) -> (Color, isize) {
    let (amount_str, color_str) = reveal.trim().split_once(' ').unwrap();

    (color_str.into(), amount_str.parse().unwrap())
}

fn is_game_possible(game: &Game, bag: &HashMap<Color, isize>) -> Option<isize> {
    let game_possible = game.reveals.iter()
        .flat_map(|reveal| reveal.iter())
        .all(|(color, number)| number <= bag.get(color).unwrap_or(&0));

    if game_possible {
        return Some(game.id);
    }

    None
}

fn get_smallest_bag(game: &Game) -> HashMap<Color, isize> {
    let mut bag = HashMap::new();
    let all_reveals = game.reveals.iter().flat_map(|reveal| reveal.iter());

    for (color, number) in all_reveals {
        let max = bag.entry(*color).or_insert(*number);
        if *max < *number {
            *max = *number;
        }
    }

    bag
}

impl Into<Color> for &str {
    fn into(self) -> Color {
        match self {
            "blue" => Color::Blue,
            "green" => Color::Green,
            "red" => Color::Red,
            _ => panic!("unknown color {}", self),
        }
    }
}

File: src/AdventOfCode.AoC2023/Solution/Day02.cs
using AdventOfCode.Core;

namespace AdventOfCode.AoC2023.Solution;

public class Day02 : ISolution
{
    public string Name => "Cube Conundrum";
    
    public string Part1(string inputFile)
    {
        var bag = new Dictionary<Color, int>
        {
            { Color.Blue, 14 },
            { Color.Green, 13 },
            { Color.Red, 12 },
        };
        
        var input = Calendar.LoadInput(inputFile);
        var games = input.Split(Environment.NewLine, StringSplitOptions.RemoveEmptyEntries)
            .Select(l => IsGamePossible(new Game(l), bag))
            .Sum();
        
        return games.ToString();
    }

    public string Part2(string inputFile)
    {
        var input = Calendar.LoadInput(inputFile);
        var games = input.Split(Environment.NewLine, StringSplitOptions.RemoveEmptyEntries)
            .Select(l => new Game(l))
            .ToArray();

        var total = games.Select(GetSmallestBag)
            .Select(bag => bag.Values.Aggregate(1, (a, b) => a * b))
            .Sum();

        return total.ToString();
    }

    private int IsGamePossible(Game game, Dictionary<Color, int> bag)
    {
        var gamePossible = game.Reveals
            .SelectMany(r => r)
            .All(r => r.Value <= bag[r.Key]);

        return gamePossible ? game.Id : 0;
    }

    private Dictionary<Color, int> GetSmallestBag(Game game)
    {
        var bag = new Dictionary<Color, int>();
        var reveals = game.Reveals.SelectMany(r => r).ToArray();

        foreach (var reveal in reveals)
        {
            if (!bag.TryGetValue(reveal.Key, out var max))
            {
                bag.Add(reveal.Key, reveal.Value);
                max = reveal.Value;
            }
            
            if (max < reveal.Value)
                bag[reveal.Key] = reveal.Value;
        }
        
        return bag;
    }
    
    [Flags]
    private enum Color
    {
        Red,
        Green,
        Blue
    }

    private class Game
    {
        public int Id;
        public List<Dictionary<Color, int>> Reveals;

        public Game(string line)
        {
            var parts = line.Split(':', count: 2);
        
            Id = int.Parse(new string(parts.First().Where(char.IsAsciiDigit).ToArray()));
            Reveals = parts.Last()
                .Split(';')
                .Select(r => r.Split(',')
                    .Select(ParseColorReveal)
                    .ToDictionary(c => c.Item1, c => c.Item2)
                )
                .ToList();
        }

        private (Color, int) ParseColorReveal(string reveal)
        {
            var parts = reveal.Trim().Split(' ', count: 2);
        
            return (Enum.Parse<Color>(parts.Last(), true), int.Parse(parts.First()));
        }
    }
}

